In [2]:
# 1. IMPORT REQUIRED LIBRARIES

import pandas as pd
import numpy as np

In [8]:
from pathlib import Path
import pandas as pd

# Load the Tickets sheet.
# This is the main fact table containing ticket-level information.
tickets = pd.read_excel(file_path, sheet_name="Tickets")

# Load the Clients sheet.
# This contains client-related attributes such as contract tier
# and SLA credit clause.
clients = pd.read_excel(file_path, sheet_name="Clients")

# Load the Agents sheet.
# This contains agent-related information such as team,
# hub, and daily capacity.
agents = pd.read_excel(file_path, sheet_name="Agents")

# Display the number of rows and columns in each dataframe.
print("Tickets:", tickets.shape)
print("Clients:", clients.shape)
print("Agents:", agents.shape)

Tickets: (3500, 16)
Clients: (80, 10)
Agents: (120, 8)


In [9]:
# Display column names for each dataframe.
print("Tickets columns:")
print(tickets.columns.tolist())

print("\nClients columns:")
print(clients.columns.tolist())

print("\nAgents columns:")
print(agents.columns.tolist())

Tickets columns:
['TicketID', 'TicketReference', 'ClientID', 'PatientRef', 'CategoryID', 'PriorityID', 'Status', 'AssignedAgentID', 'CreatedAt', 'FirstResponseAt', 'ResolvedAt', 'SLADueAt', 'SLABreached', 'Channel', 'Description', 'ResolutionNotes']

Clients columns:
['ClientID', 'ClientName', 'ClientType', 'ContractTier', 'AccountManagerID', 'Region', 'ContractStartDate', 'ContractEndDate', 'SLACreditClause', 'IsActive']

Agents columns:
['AgentID', 'FullName', 'TeamID', 'Role', 'Specialisms', 'Hub', 'DailyCapacity', 'IsActive']


In [10]:
# Display data types and non-null counts for each dataframe.
print("Tickets info:")
tickets.info()

print("\nClients info:")
clients.info()

print("\nAgents info:")
agents.info()

Tickets info:
<class 'pandas.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   TicketID         3500 non-null   int64
 1   TicketReference  3500 non-null   str  
 2   ClientID         3500 non-null   int64
 3   PatientRef       3500 non-null   str  
 4   CategoryID       3500 non-null   int64
 5   PriorityID       3500 non-null   int64
 6   Status           3500 non-null   str  
 7   AssignedAgentID  3500 non-null   int64
 8   CreatedAt        3500 non-null   str  
 9   FirstResponseAt  3454 non-null   str  
 10  ResolvedAt       3301 non-null   str  
 11  SLADueAt         3500 non-null   str  
 12  SLABreached      3500 non-null   int64
 13  Channel          3500 non-null   str  
 14  Description      3500 non-null   str  
 15  ResolutionNotes  3301 non-null   str  
dtypes: int64(6), str(10)
memory usage: 437.6 KB

Clients info:
<class 'pandas.DataFrame'>
RangeIn

In [11]:
# Calculate missing values for the Tickets dataframe.
tickets_missing = pd.DataFrame({
    "Missing_Count": tickets.isnull().sum(),
    "Missing_Percentage": (tickets.isnull().mean() * 100).round(2)
})

print("Missing values - Tickets")
display(tickets_missing)

# Calculate missing values for the Clients dataframe.
clients_missing = pd.DataFrame({
    "Missing_Count": clients.isnull().sum(),
    "Missing_Percentage": (clients.isnull().mean() * 100).round(2)
})

print("Missing values - Clients")
display(clients_missing)

# Calculate missing values for the Agents dataframe.
agents_missing = pd.DataFrame({
    "Missing_Count": agents.isnull().sum(),
    "Missing_Percentage": (agents.isnull().mean() * 100).round(2)
})

print("Missing values - Agents")
display(agents_missing)

Missing values - Tickets


,Missing_Count,Missing_Percentage
TicketID,0,0.00
TicketReference,0,0.00
ClientID,0,0.00
PatientRef,0,0.00
CategoryID,0,0.00
PriorityID,0,0.00
Status,0,0.00
AssignedAgentID,0,0.00
CreatedAt,0,0.00
FirstResponseAt,46,1.31


Missing values - Clients


,Missing_Count,Missing_Percentage
ClientID,0,0.0
ClientName,0,0.0
ClientType,0,0.0
ContractTier,0,0.0
AccountManagerID,0,0.0
Region,0,0.0
ContractStartDate,0,0.0
ContractEndDate,0,0.0
SLACreditClause,0,0.0
IsActive,0,0.0


Missing values - Agents


,Missing_Count,Missing_Percentage
AgentID,0,0.0
FullName,0,0.0
TeamID,0,0.0
Role,0,0.0
Specialisms,0,0.0
Hub,0,0.0
DailyCapacity,0,0.0
IsActive,0,0.0


In [12]:
# Check for completely duplicated rows in each dataframe.
print("Duplicate rows:")
print("Tickets:", tickets.duplicated().sum())
print("Clients:", clients.duplicated().sum())
print("Agents:", agents.duplicated().sum())

# Check whether the primary key columns contain duplicates.
print("\nDuplicate primary keys:")
print("TicketID:", tickets["TicketID"].duplicated().sum())
print("ClientID:", clients["ClientID"].duplicated().sum())
print("AgentID:", agents["AgentID"].duplicated().sum())

Duplicate rows:
Tickets: 0
Clients: 0
Agents: 0

Duplicate primary keys:
TicketID: 0
ClientID: 0
AgentID: 0


In [13]:
# Check the number of unique values in each column.
# This helps identify categorical variables and high-cardinality columns.

print("Tickets - Unique values:")
print(tickets.nunique())

print("\nClients - Unique values:")
print(clients.nunique())

print("\nAgents - Unique values:")
print(agents.nunique())

Tickets - Unique values:
TicketID           3500
TicketReference    3500
ClientID             72
PatientRef         3441
CategoryID            5
PriorityID            4
Status                6
AssignedAgentID     114
CreatedAt          3496
FirstResponseAt    3450
ResolvedAt         3298
SLADueAt           3495
SLABreached           2
Channel               4
Description        3214
ResolutionNotes     284
dtype: int64

Clients - Unique values:
ClientID             80
ClientName           68
ClientType            6
ContractTier          3
AccountManagerID     27
Region                6
ContractStartDate    61
ContractEndDate      80
SLACreditClause       2
IsActive              2
dtype: int64

Agents - Unique values:
AgentID          120
FullName         119
TeamID             4
Role               4
Specialisms       62
Hub                4
DailyCapacity     20
IsActive           2
dtype: int64


In [14]:
# Examine the distinct values and frequency of key categorical columns
# in the Tickets dataframe.

categorical_columns = [
    "Status",
    "PriorityID",
    "CategoryID",
    "Channel",
    "SLABreached"
]

for column in categorical_columns:
    print(f"\n{column}:")
    print(tickets[column].value_counts(dropna=False))


Status:
Status
Closed            1984
Resolved          1317
Escalated           57
In Progress         52
Open                46
Pending Client      44
Name: count, dtype: int64

PriorityID:
PriorityID
3    1981
2     833
4     428
1     258
Name: count, dtype: int64

CategoryID:
CategoryID
1    1200
2     873
3     706
5     415
4     306
Name: count, dtype: int64

Channel:
Channel
Email     1221
Portal     977
Phone      905
Form       397
Name: count, dtype: int64

SLABreached:
SLABreached
0    2746
1     754
Name: count, dtype: int64


In [15]:
# Check whether every ClientID in Tickets exists in the Clients table.
invalid_clients = tickets[
    ~tickets["ClientID"].isin(clients["ClientID"])
]

# Check whether every AssignedAgentID in Tickets exists in the Agents table.
invalid_agents = tickets[
    ~tickets["AssignedAgentID"].isin(agents["AgentID"])
]

print("Tickets with invalid ClientID:", len(invalid_clients))
print("Tickets with invalid AssignedAgentID:", len(invalid_agents))

Tickets with invalid ClientID: 0
Tickets with invalid AssignedAgentID: 0


In [16]:
# Display the distribution of important categorical variables
# as both counts and percentages.

categorical_columns = [
    "Status",
    "PriorityID",
    "CategoryID",
    "Channel",
    "SLABreached"
]

for column in categorical_columns:
    
    # Count the number of records in each category.
    counts = tickets[column].value_counts(dropna=False)
    
    # Calculate the percentage represented by each category.
    percentages = (
        tickets[column]
        .value_counts(normalize=True, dropna=False)
        .mul(100)
        .round(2)
    )
    
    # Combine counts and percentages into one table.
    distribution = pd.DataFrame({
        "Count": counts,
        "Percentage": percentages
    })
    
    print(f"\n{column} Distribution:")
    display(distribution)


Status Distribution:


,Count,Percentage
Status,,
Closed,1984,56.69
Resolved,1317,37.63
Escalated,57,1.63
In Progress,52,1.49
Open,46,1.31
Pending Client,44,1.26



PriorityID Distribution:


,Count,Percentage
PriorityID,,
3,1981,56.60
2,833,23.80
4,428,12.23
1,258,7.37



CategoryID Distribution:


,Count,Percentage
CategoryID,,
1,1200,34.29
2,873,24.94
3,706,20.17
5,415,11.86
4,306,8.74



Channel Distribution:


,Count,Percentage
Channel,,
Email,1221,34.89
Portal,977,27.91
Phone,905,25.86
Form,397,11.34



SLABreached Distribution:


,Count,Percentage
SLABreached,,
0,2746,78.46
1,754,21.54


In [17]:
# List the date columns that need to be checked.
date_columns = [
    "CreatedAt",
    "FirstResponseAt",
    "ResolvedAt",
    "SLADueAt"
]

# Profile each date column without modifying the original dataframe.
for column in date_columns:
    
    # Convert the column temporarily to datetime.
    # Invalid date values will become NaT (Not a Time).
    converted_dates = pd.to_datetime(
        tickets[column],
        errors="coerce"
    )
    
    # Count values that could not be converted,
    # excluding values that were already missing.
    invalid_dates = (
        converted_dates.isna() & tickets[column].notna()
    ).sum()
    
    print(f"\n{column}:")
    print("Invalid dates:", invalid_dates)
    print("Earliest date:", converted_dates.min())
    print("Latest date:", converted_dates.max())


CreatedAt:
Invalid dates: 0
Earliest date: 2023-01-01 10:54:00
Latest date: 2025-03-31 07:56:00

FirstResponseAt:
Invalid dates: 0
Earliest date: 2023-01-01 14:16:00
Latest date: 2025-03-31 09:21:00

ResolvedAt:
Invalid dates: 0
Earliest date: 2023-01-02 03:35:00
Latest date: 2025-04-04 11:02:00

SLADueAt:
Invalid dates: 0
Earliest date: 2023-01-02 08:04:00
Latest date: 2025-04-02 19:28:00


In [18]:
# Temporarily convert the date columns to datetime
# so that we can check the logical order of events.
created = pd.to_datetime(tickets["CreatedAt"], errors="coerce")
first_response = pd.to_datetime(tickets["FirstResponseAt"], errors="coerce")
resolved = pd.to_datetime(tickets["ResolvedAt"], errors="coerce")
sla_due = pd.to_datetime(tickets["SLADueAt"], errors="coerce")

# Check for impossible or suspicious date relationships.
response_before_creation = (
    first_response.notna() & (first_response < created)
).sum()

resolution_before_creation = (
    resolved.notna() & (resolved < created)
).sum()

sla_before_creation = (
    sla_due.notna() & (sla_due < created)
).sum()

print("First response before ticket creation:", response_before_creation)
print("Resolution before ticket creation:", resolution_before_creation)
print("SLA due before ticket creation:", sla_before_creation)

First response before ticket creation: 0
Resolution before ticket creation: 0
SLA due before ticket creation: 0


In [19]:
# Profile important numerical columns.

print("Tickets numerical summary:")
display(
    tickets[
        ["CategoryID", "PriorityID", "SLABreached"]
    ].describe()
)

print("\nClients numerical summary:")
display(
    clients[
        ["SLACreditClause", "IsActive"]
    ].describe()
)

print("\nAgents numerical summary:")
display(
    agents[
        ["TeamID", "DailyCapacity", "IsActive"]
    ].describe()
)

Tickets numerical summary:


,CategoryID,PriorityID,SLABreached
count,3500.000000,3500.000000,3500.000000
mean,2.389429,2.736857,0.215429
std,1.345471,0.765550,0.411178
min,1.000000,1.000000,0.000000
25%,1.000000,2.000000,0.000000
50%,2.000000,3.000000,0.000000
75%,3.000000,3.000000,0.000000
max,5.000000,4.000000,1.000000



Clients numerical summary:


,SLACreditClause,IsActive
count,80.000000,80.000000
mean,0.387500,0.900000
std,0.490253,0.301893
min,0.000000,0.000000
25%,0.000000,1.000000
50%,0.000000,1.000000
75%,1.000000,1.000000
max,1.000000,1.000000



Agents numerical summary:


,TeamID,DailyCapacity,IsActive
count,120.000000,120.000000,120.000000
mean,4.150000,20.925000,0.950000
std,2.655737,4.692498,0.218859
min,1.000000,6.000000,0.000000
25%,1.000000,19.000000,1.000000
50%,4.000000,21.000000,1.000000
75%,6.000000,24.250000,1.000000
max,8.000000,28.000000,1.000000


In [20]:
# Check the actual values used by binary fields.
# These fields should normally contain only 0 and 1.

print("SLABreached values:")
print(tickets["SLABreached"].value_counts())

print("\nClient IsActive values:")
print(clients["IsActive"].value_counts())

print("\nSLA Credit Clause values:")
print(clients["SLACreditClause"].value_counts())

print("\nAgent IsActive values:")
print(agents["IsActive"].value_counts())

SLABreached values:
SLABreached
0    2746
1     754
Name: count, dtype: int64

Client IsActive values:
IsActive
1    72
0     8
Name: count, dtype: int64

SLA Credit Clause values:
SLACreditClause
0    49
1    31
Name: count, dtype: int64

Agent IsActive values:
IsActive
1    114
0      6
Name: count, dtype: int64


In [21]:
# Identify inactive clients.
inactive_clients = clients.loc[
    clients["IsActive"] == 0,
    "ClientID"
]

# Check whether any tickets belong to inactive clients.
tickets_inactive_clients = tickets[
    tickets["ClientID"].isin(inactive_clients)
]

# Identify inactive agents.
inactive_agents = agents.loc[
    agents["IsActive"] == 0,
    "AgentID"
]

# Check whether any tickets are assigned to inactive agents.
tickets_inactive_agents = tickets[
    tickets["AssignedAgentID"].isin(inactive_agents)
]

print(
    "Tickets belonging to inactive clients:",
    len(tickets_inactive_clients)
)

print(
    "Tickets assigned to inactive agents:",
    len(tickets_inactive_agents)
)

Tickets belonging to inactive clients: 0
Tickets assigned to inactive agents: 0


In [22]:
# Examine the statistical distribution of agent daily capacity.
print("Daily Capacity Summary:")
print(agents["DailyCapacity"].describe())

# Display the different capacity values and
# the number of agents with each capacity.
print("\nDaily Capacity Distribution:")
print(
    agents["DailyCapacity"]
    .value_counts()
    .sort_index()
)

Daily Capacity Summary:
count    120.000000
mean      20.925000
std        4.692498
min        6.000000
25%       19.000000
50%       21.000000
75%       24.250000
max       28.000000
Name: DailyCapacity, dtype: float64

Daily Capacity Distribution:
DailyCapacity
6      1
8      2
10     1
11     1
13     3
14     4
15     2
16     3
17     3
18     9
19    17
20    10
21    12
22     6
23     7
24     9
25     7
26     7
27     7
28     9
Name: count, dtype: int64


In [23]:
# Display agents with the lowest daily capacities.
# This helps determine whether low capacity values are associated
# with particular roles, teams or hubs.

low_capacity_agents = (
    agents[
        ["AgentID", "Role", "TeamID", "Hub", "DailyCapacity"]
    ]
    .sort_values("DailyCapacity")
    .head(10)
)

display(low_capacity_agents)

,AgentID,Role,TeamID,Hub,DailyCapacity
30,31,Operations Manager,1,Manchester,6
31,32,Team Lead,1,Manchester,8
60,61,Team Lead,4,Leeds,8
91,92,Team Lead,6,Birmingham,10
1,2,Team Lead,1,Manchester,11
44,45,Team Lead,4,Leeds,13
104,105,Team Lead,8,Edinburgh,13
118,119,Team Lead,8,Edinburgh,13
77,78,Team Lead,6,Birmingham,14
39,40,Team Lead,1,Manchester,14


In [24]:
# Compare daily capacity across different agent roles.
# This helps identify whether capacity differences are
# related to the type of work performed by each role.

capacity_by_role = (
    agents
    .groupby("Role")["DailyCapacity"]
    .agg(["count", "min", "mean", "median", "max"])
    .round(2)
)

display(capacity_by_role)

,count,min,mean,median,max
Role,,,,,
Agent,92,18,22.75,22.5,28
Operations Manager,1,6,6.00,6.0,6
Senior Agent,18,14,17.11,17.0,20
Team Lead,9,8,11.56,13.0,14


### Daily Capacity Profiling Findings

- DailyCapacity ranges from 6 to 28 tickets per day.
- Lower capacity values are associated with senior/supervisory roles.
- Agents have the highest average capacity (22.75).
- Senior Agents average 17.11.
- Team Leads average 11.56.
- The minimum capacity of 6 belongs to the Operations Manager.
- Therefore, the low capacity values appear role-related rather than obvious data-quality errors.
- No DailyCapacity values will be removed or modified during profiling.

In [25]:
# Profile the client contract features that will later
# be considered for the SLA breach prediction model.

client_features = [
    "ContractTier",
    "SLACreditClause"
]

for column in client_features:

    # Count clients in each category.
    counts = clients[column].value_counts(dropna=False)

    # Calculate the percentage of clients in each category.
    percentages = (
        clients[column]
        .value_counts(normalize=True, dropna=False)
        .mul(100)
        .round(2)
    )

    # Combine the counts and percentages.
    distribution = pd.DataFrame({
        "Count": counts,
        "Percentage": percentages
    })

    print(f"\n{column} Distribution:")
    display(distribution)


ContractTier Distribution:


,Count,Percentage
ContractTier,,
Standard,37,46.25
Premium,29,36.25
Enterprise,14,17.50



SLACreditClause Distribution:


,Count,Percentage
SLACreditClause,,
0,49,61.25
1,31,38.75


### Client Feature Profiling Findings

- ContractTier contains three categories: Standard, Premium and Enterprise.
- Standard is the most common contract tier (46.25%), followed by Premium (36.25%) and Enterprise (17.50%).
- 38.75% of clients have an SLA credit clause, while 61.25% do not.
- No missing or unexpected values were identified in ContractTier or SLACreditClause.
- Both variables appear suitable for further analysis and potential use as model features.

In [26]:
# Join client contract information to the ticket data
# for profiling purposes only.

tickets_clients = tickets.merge(
    clients[
        ["ClientID", "ContractTier", "SLACreditClause"]
    ],
    on="ClientID",
    how="left"
)

# Confirm that the join did not change the number of tickets.
print("Tickets before join:", len(tickets))
print("Tickets after join:", len(tickets_clients))

Tickets before join: 3500
Tickets after join: 3500


In [27]:
# Calculate SLA breach rate by Contract Tier.
# The mean of a binary 0/1 variable gives the proportion of breached tickets.

breach_by_contract = (
    tickets_clients
    .groupby("ContractTier")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate to a percentage.
breach_by_contract["Breach_Rate"] = (
    breach_by_contract["Breach_Rate"] * 100
).round(2)

display(breach_by_contract)

,Total_Tickets,Breached_Tickets,Breach_Rate
ContractTier,,,
Enterprise,543,113,20.81
Premium,1295,277,21.39
Standard,1662,364,21.90


In [28]:
# Calculate SLA breach rate based on whether
# the client has an SLA credit clause.

breach_by_credit_clause = (
    tickets_clients
    .groupby("SLACreditClause")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate to a percentage.
breach_by_credit_clause["Breach_Rate"] = (
    breach_by_credit_clause["Breach_Rate"] * 100
).round(2)

display(breach_by_credit_clause)

,Total_Tickets,Breached_Tickets,Breach_Rate
SLACreditClause,,,
0,2189,475,21.70
1,1311,279,21.28


### SLA Breach by Client Characteristics

- SLA breach rates are relatively similar across all three contract tiers.
- Standard clients have the highest observed breach rate at 21.90%, followed by Premium at 21.39% and Enterprise at 20.81%.
- Clients without an SLA credit clause have a breach rate of 21.70%, compared with 21.28% for clients with an SLA credit clause.
- The differences are small at the individual-variable level.
- These variables should not be assumed to be strong predictors based on profiling alone, but they can still be evaluated alongside other features during modelling.

In [29]:
# Calculate SLA breach rate for each ticket priority.

breach_by_priority = (
    tickets
    .groupby("PriorityID")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate from a proportion to percentage.
breach_by_priority["Breach_Rate"] = (
    breach_by_priority["Breach_Rate"] * 100
).round(2)

display(breach_by_priority)

,Total_Tickets,Breached_Tickets,Breach_Rate
PriorityID,,,
1,258,74,28.68
2,833,168,20.17
3,1981,426,21.50
4,428,86,20.09


### SLA Breach by Ticket Priority

- Priority 1 tickets have the highest observed SLA breach rate at 28.68%.
- Priority 2, 3 and 4 tickets have breach rates of 20.17%, 21.50% and 20.09% respectively.
- Priority 1 therefore shows a noticeably higher breach rate than the other priority groups.
- This suggests that PriorityID may contain useful predictive information for the SLA breach model.
- The relationship identified during profiling is associative and should not be interpreted as causal.

In [30]:
# Calculate SLA breach rate for each ticket category.

breach_by_category = (
    tickets
    .groupby("CategoryID")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate to percentage.
breach_by_category["Breach_Rate"] = (
    breach_by_category["Breach_Rate"] * 100
).round(2)

display(breach_by_category)

,Total_Tickets,Breached_Tickets,Breach_Rate
CategoryID,,,
1,1200,245,20.42
2,873,180,20.62
3,706,154,21.81
4,306,74,24.18
5,415,101,24.34


### SLA Breach by Ticket Category

- Category 5 has the highest observed SLA breach rate at 24.34%.
- Category 4 follows closely at 24.18%.
- Categories 1 and 2 have lower breach rates of 20.42% and 20.62%.
- Category 3 has a breach rate of 21.81%.
- The variation across categories suggests that CategoryID may provide useful predictive information for the SLA breach model.
- These results show association only and do not establish that ticket category causes SLA breaches.

In [31]:
# Calculate SLA breach rate for each ticket channel.

breach_by_channel = (
    tickets
    .groupby("Channel")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate from a proportion to a percentage.
breach_by_channel["Breach_Rate"] = (
    breach_by_channel["Breach_Rate"] * 100
).round(2)

# Display the result.
display(breach_by_channel)

,Total_Tickets,Breached_Tickets,Breach_Rate
Channel,,,
Email,1221,246,20.15
Form,397,89,22.42
Phone,905,216,23.87
Portal,977,203,20.78


### SLA Breach by Ticket Channel

- Phone tickets have the highest observed SLA breach rate at 23.87%.
- Form tickets have a breach rate of 22.42%.
- Portal tickets have a breach rate of 20.78%.
- Email tickets have the lowest observed breach rate at 20.15%.
- The differences suggest that Channel may provide useful information for SLA breach prediction.
- These results show association and should not be interpreted as evidence that the communication channel causes SLA breaches.

In [32]:
# Join agent information to the ticket data for profiling purposes.

tickets_agents = tickets.merge(
    agents[
        ["AgentID", "Hub", "TeamID", "DailyCapacity"]
    ],
    left_on="AssignedAgentID",
    right_on="AgentID",
    how="left"
)

# Confirm that the join has not created or removed ticket records.
print("Tickets before join:", len(tickets))
print("Tickets after join:", len(tickets_agents))

Tickets before join: 3500
Tickets after join: 3500


In [33]:
# Calculate SLA breach rate for each operational hub.

breach_by_hub = (
    tickets_agents
    .groupby("Hub")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate from a proportion to a percentage.
breach_by_hub["Breach_Rate"] = (
    breach_by_hub["Breach_Rate"] * 100
).round(2)

# Display the results.
display(breach_by_hub)

,Total_Tickets,Breached_Tickets,Breach_Rate
Hub,,,
Birmingham,736,173,23.51
Edinburgh,817,156,19.09
Leeds,601,113,18.80
Manchester,1346,312,23.18


### SLA Breach by Operational Hub

- Birmingham has the highest observed SLA breach rate at 23.51%.
- Manchester follows closely at 23.18%.
- Edinburgh has a breach rate of 19.09%.
- Leeds has the lowest observed breach rate at 18.80%.
- The difference between the highest and lowest hub breach rates is approximately 4.7 percentage points.
- This suggests that Hub may provide useful information for SLA breach prediction.
- The observed relationship is associative and does not establish that operational hub causes SLA breaches.

In [34]:
# Calculate SLA breach rate for each operational team.

breach_by_team = (
    tickets_agents
    .groupby("TeamID")["SLABreached"]
    .agg(["count", "sum", "mean"])
    .rename(columns={
        "count": "Total_Tickets",
        "sum": "Breached_Tickets",
        "mean": "Breach_Rate"
    })
)

# Convert the breach rate to a percentage.
breach_by_team["Breach_Rate"] = (
    breach_by_team["Breach_Rate"] * 100
).round(2)

# Sort from highest to lowest breach rate
# to make higher-risk teams easier to identify.
breach_by_team = breach_by_team.sort_values(
    "Breach_Rate",
    ascending=False
)

display(breach_by_team)

,Total_Tickets,Breached_Tickets,Breach_Rate
TeamID,,,
6,736,173,23.51
1,1346,312,23.18
8,817,156,19.09
4,601,113,18.80


In [35]:
# Check the relationship between Hub and TeamID.
# This helps determine whether the two variables
# contain the same or overlapping information.

hub_team_relationship = pd.crosstab(
    tickets_agents["Hub"],
    tickets_agents["TeamID"]
)

display(hub_team_relationship)

TeamID,1,4,6,8
Hub,,,,
Birmingham,0,0,736,0
Edinburgh,0,0,0,817
Leeds,0,601,0,0
Manchester,1346,0,0,0


In [36]:
# Count how many different TeamIDs are associated
# with each operational hub.

teams_per_hub = (
    tickets_agents
    .groupby("Hub")["TeamID"]
    .nunique()
)

print("Number of teams represented in tickets per hub:")
print(teams_per_hub)

Number of teams represented in tickets per hub:
Hub
Birmingham    1
Edinburgh     1
Leeds         1
Manchester    1
Name: TeamID, dtype: int64


### Hub and Team Relationship

- Each operational hub has only one TeamID represented in the ticket dataset.
- Manchester maps to TeamID 1.
- Leeds maps to TeamID 4.
- Birmingham maps to TeamID 6.
- Edinburgh maps to TeamID 8.
- Consequently, Hub and TeamID produce identical SLA breach-rate patterns in the current ticket data.
- The two variables may therefore contain redundant predictive information.
- No variable is removed during profiling; this will be considered during feature selection and modelling.

In [37]:
# Convert CreatedAt temporarily to datetime for profiling.
# This does not change the original CreatedAt column.
created_at = pd.to_datetime(
    tickets["CreatedAt"],
    errors="coerce"
)

# Extract the hour when each ticket was created.
created_hour = created_at.dt.hour

# Extract the day of the week.
# Monday = 0 and Sunday = 6.
created_day = created_at.dt.dayofweek

# Display ticket volume by hour of day.
print("Tickets by creation hour:")
display(
    created_hour
    .value_counts()
    .sort_index()
)

# Display ticket volume by day of week.
print("\nTickets by day of week:")
display(
    created_day
    .value_counts()
    .sort_index()
)

Tickets by creation hour:


CreatedAt
0     130
1     139
2     148
3     159
4     136
5     143
6     130
7     163
8     151
9     153
10    138
11    150
12    168
13    155
14    114
15    152
16    134
17    157
18    154
19    140
20    147
21    138
22    130
23    171
Name: count, dtype: int64


Tickets by day of week:


CreatedAt
0    468
1    493
2    512
3    473
4    519
5    522
6    513
Name: count, dtype: int64

### Ticket Creation Time Profiling

- Ticket creation occurs across all 24 hours of the day.
- Hourly ticket volumes are relatively evenly distributed, with no major concentration at a specific hour.
- Hour 23 has the highest observed volume with 171 tickets, while hour 14 has the lowest with 114 tickets.
- Ticket volumes are also relatively evenly distributed across the days of the week.
- Saturday has the highest volume with 522 tickets, while Monday has the lowest with 468 tickets.
- Creation time may still provide useful information for SLA breach prediction and will be investigated further during EDA and feature engineering.

## Data Profiling Summary

### Dataset Overview
- Tickets: 3,500 rows and 16 columns.
- Clients: 80 rows and 10 columns.
- Agents: 120 rows and 8 columns.

### Data Completeness
- Clients and Agents contain no missing values.
- Tickets contains 46 missing FirstResponseAt values (1.31%).
- Tickets contains 199 missing ResolvedAt values (5.69%).
- Tickets contains 199 missing ResolutionNotes values (5.69%).
- The matching missing counts for ResolvedAt and ResolutionNotes may be associated with unresolved tickets and should be investigated during data cleaning.

### Duplicate and Key Checks
- No completely duplicated rows were identified.
- No duplicate TicketID, ClientID or AgentID values were identified.
- All ClientIDs referenced by Tickets exist in Clients.
- All AssignedAgentIDs referenced by Tickets exist in Agents.
- No tickets reference inactive clients or inactive agents.

### Date and Time Quality
- CreatedAt, FirstResponseAt, ResolvedAt and SLADueAt contain no invalid date values.
- No first responses occur before ticket creation.
- No resolutions occur before ticket creation.
- No SLA due dates occur before ticket creation.
- Date columns are currently stored as strings and should be converted to datetime during data cleaning.

### SLA Breach Target
- 2,746 tickets (78.46%) did not breach SLA.
- 754 tickets (21.54%) breached SLA.
- The target therefore shows some class imbalance, which should be considered during modelling and evaluation.

### Potential Predictive Patterns
- Priority 1 has the highest breach rate at 28.68%.
- Categories 4 and 5 have higher breach rates than the other ticket categories.
- Phone tickets have the highest channel breach rate at 23.87%.
- Birmingham and Manchester have higher observed breach rates than Leeds and Edinburgh.
- ContractTier and SLACreditClause show relatively small differences in breach rates.

### Agent Capacity
- DailyCapacity ranges from 6 to 28 tickets per day.
- Lower capacities are associated with supervisory roles such as Team Lead and Operations Manager.
- The low capacity values therefore appear role-related rather than obvious data-quality errors.

### Hub and Team Relationship
- Each hub has only one TeamID represented in the ticket dataset.
- Hub and TeamID therefore encode overlapping information in the current modelling data.
- This potential redundancy should be considered later during feature selection.

### Ticket Creation Patterns
- Tickets are created throughout all 24 hours.
- Ticket volume is relatively evenly distributed across the days of the week.
- Creation hour and day of week may be investigated further during EDA and feature engineering.

### Next Step
The profiling stage identified a generally well-structured dataset with a small number of missing values and some datatype issues. No records have been modified during profiling. The next stage is Data Cleaning, where identified quality issues will be investigated and handled systematically.